### PR TITLE
Added tests for ep_skip_query_integration and edited ep_skip_query_integration from tests/feedback

### PR DESCRIPTION
**Pull request by brandon-m-skinner**
## Description

We were missing some important tests for query integration. So I added them.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).

## Steps to Test

1. Check out PR.
2. `./bin/phpunit-docker.sh tests/search/test-class-search.php`

### CODE REVIEW

File: tests/search/test-class-search.php [hunk 1]
```diff
@@ -862,4 +862,61 @@ public function test__vip_search_filter__ep_allow_post_content_filtered_index()
 
 		$this->assertFalse( $enabled );
 	}
+
+	/*
+	 * Ensure the various options, constants, and rate limiting affect query integration in the right ways
+	 */
+	public function test__ep_skip_query_integration() {
```
Comment (mjangda) [inline tests/search/test-class-search.php:869]: Minor: Should we split this off into individual tests? I personally find large tests like this hard to follow and they also tend to be fragile (if we change one aspect of how the query integration stuff works, the test may need to ba completely re-written).

File: tests/search/test-class-search.php [hunk 2]
```diff
@@ -862,4 +862,61 @@ public function test__vip_search_filter__ep_allow_post_content_filtered_index()
...
+	}
+
+	/**
+	 * Ensure the incrementor for tracking request counts behaves properly
+	 */
+	public function test__query_count_incr() {
+		$es = new \Automattic\VIP\Search\Search();
+
+		// Reset cache key
+		wp_cache_delete( $es::QUERY_COUNT_CACHE_KEY, $es::QUERY_COUNT_CACHE_GROUP );
```
[the remark this answers is not in the document]
Comment (brandon-m-skinner) [reply tests/search/test-class-search.php:914]: Probably not right now, but it's the safe bet. Nothing is worse than having tests you didn't touch fail when you unintentionally cause side effects.

File: search/includes/classes/class-search.php [hunk 3]
```diff
@@ -664,7 +665,7 @@ public function filter__epwr_boost_mode( $boost_mode, $formatted_args, $args ) {
 	/*
 	 * Increment the number of queries that have been passed through VIP Search
 	 */
-	private static function query_count_incr() {
+	public static function query_count_incr() {
```
[the remark this answers is not in the document]
Comment (brandon-m-skinner) [reply search/includes/classes/class-search.php:668]: Interesting, I'll give that a shot.

File: search/includes/classes/class-search.php [hunk 4]
```diff
@@ -339,6 +339,15 @@ public function filter__ep_feature_active( $active, $feature_settings, $feature
 	 * constant should be set to `true`, which will enable query integration for all requests
 	 */
 	public static function ep_skip_query_integration( $skip, $query = null ) {
+		// If the query count has exceeded the maximum
+		//     Only allow half of the queries to use VIP Search
+		if ( self::query_count_incr() > self::$max_query_count ) {
```
Comment (mjangda) [inline search/includes/classes/class-search.php:344]: One thought I had: should we only trigger this when we determine `$skip === false`? In contexts where we return `true`, and ES doesn't kick in, we're doing unnecessary memcache operations and incorrectly incrementing rate limit counts.

Comment (brandon-m-skinner) [reply search/includes/classes/class-search.php:344]: True, the skip check should be at the top 👍

Comment (nickdaugherty) [reply search/includes/classes/class-search.php:356]: I think what Mo was getting at is it would be better to run all the logic in this function first (remove the early return on the `$_GET['es']` check), and _then_ only rate limit if `$skip === false`.

Currently, we check the counter even if the rest of the filter ultimately returns `true` and skips the integration (if the option or constant is not set).

Comment (nickdaugherty) [reply search/includes/classes/class-search.php:356]: We could add the rate limit as a separate late running (PHP_INT_MAX) filter on `es_skip_query_integration` to solve those concerns - let this filter callback run normally, then our rate limit filter callback runs at the very end, disabling the query integration if we're rate limited (and query integration had been enabled via other filters).

Comment (mjangda) [reply search/includes/classes/class-search.php:356]: > We could add the rate limit as a separate late running (PHP_INT_MAX) filter on es_skip_query_integration to solve those concerns

Ah, that a great idea. Much cleaner and easier to disable / detangle if we ever need to.

Comment (brandon-m-skinner) [reply search/includes/classes/class-search.php:356]: Sounds good to me :+1:

File: search/includes/classes/class-search.php [hunk 5]
```diff
@@ -395,6 +390,34 @@ public static function ep_skip_query_integration( $skip, $query = null ) {
...
+	 * @return bool new value of ep_skip_query_integration
+	 */
+	public static function rate_limit_ep_query_integration( $skip ) {
+		// Honor previous filters that skip query integration
+		if ( $skip ) {
+			return true;
+		}
+
+		// If the query count has exceeded the maximum
+		//     Only allow half of the queries to use VIP Search
```
Comment (nickdaugherty) [inline search/includes/classes/class-search.php:410]:
```suggestion
		// Only allow half of the queries to use VIP Search
```

**Comment by brandon-m-skinner** [comment]
r913-stacks
